File: Recursion/124. Binary Tree Maximum Path Sum.cpp
/*
124. Binary Tree Maximum Path Sum
Given a non-empty binary tree, find the maximum path sum.

For this problem, a path is defined as any sequence of nodes from some starting node to any node in the tree along the parent-child connections. The path must contain at least one node and does not need to go through the root.

Example 1:

Input: [1,2,3]

       1
      / \
     2   3

Output: 6
Example 2:

Input: [-10,9,20,null,null,15,7]

   -10
   / \
  9  20
    /  \
   15   7

Output: 42


*/

/*

  -10                       42                        
   / \                    /   \
  9  20                   9  20+15+7 = 42
    /  \                      / \
   15   7                    15  7
   先找左面的最大的，再找右面最大的，然后结合看是 l大, or r大, or l+root+r 大，or root 大， 再跟max比

*/



/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int maxval = numeric_limits<int>::min();
        helper(root,maxval);
        return maxval;
    }
    
    int helper(TreeNode *root, int &maxval){
        if(!root) return 0;
        int left = helper(root->left, maxval);
        int right = helper(root->right,maxval);
        int child = root->val;
        child = max(child, root->val+left);
        child = max(child, root->val+right);
        if(child>maxval) maxval = child;
        if(root->val+left+right>maxval) maxval = root->val+left+right;
        return child;
    }
};


class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int maxval = numeric_limits<int>::min();
        helper(root,maxval);
        return maxval;
    }
    
    int helper(TreeNode *root, int &maxval){
        if(!root) return 0;
        int left = max(0,helper(root->left, maxval));
        int right = max(0,helper(root->right,maxval));
        maxval = max(maxval,left + right + root->val);
        return max(left, right)+root->val;
    }
};



class Solution {
public:
/*    int dfs(TreeNode * root, int & maxSum, list<TreeNode*>&cur, list<TreeNode*>&maxRet) {
        if (!root) return 0;
        
    }
*/
    int dfs(TreeNode*root, int &maxSum) {
        if (!root) return 0;
        int left = dfs(root->left, maxSum);
        int right = dfs(root->right, maxSum);
        maxSum = max({maxSum, root->val, root->val + left, root->val + right, left + right + root->val});
        return max({root->val, root->val + left, root->val + right});
    }
    int maxPathSum(TreeNode* root) {
        int maxSum = INT_MIN;
        dfs(root, maxSum);
        return maxSum;
    }
};

File: Recursion/124. Binary Tree Maximum Path Sum.py
"""
124. Binary Tree Maximum Path Sum

Example 1:

Input: [1,2,3]

       1
      / \
     2   3

Output: 6
Example 2:

Input: [-10,9,20,null,null,15,7]

   -10
   / \
  9  20
    /  \
   15   7

"""

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        val = [float("-inf")]
        def dfs(root):
            if root is None: return 0
            left =  dfs(root.left)
            right = dfs(root.right)
            cur = max(left+right+root.val, root.val, root.val+left, root.val+right)
            if cur > val[0]:
                val[0] = cur
            return max(max(left,right)+root.val,root.val)
        dfs(root)
        return val[0]  
       
class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        def helper(root):
            if not root:
                return -float('inf'), -float('inf')
            left, ltot = helper(root.left)
            right, rtot = helper(root.right)
            nxt = max(left + root.val, right + root.val, root.val)
            return nxt, max(ltot, rtot, nxt, left + right + root.val)
        return max(helper(root))
       
class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        def maxsums(node):
            if not node:
                return [-2**31] * 2
            left = maxsums(node.left)
            right = maxsums(node.right)
            return [node.val + max(left[0], right[0], 0),
                    max(left + right + [node.val + left[0] + right[0]])]
        return max(maxsums(root))
